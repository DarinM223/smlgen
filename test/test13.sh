../smlgen --test test13.sml t:sceh
if cmp --silent test13.sml.test test13.sml.expected; then
  echo "test13 succeeded"
else
  echo "test13 failed"
  echo "Diff:"
  diff test13.sml.test test13.sml.expected
  exit 1
fi