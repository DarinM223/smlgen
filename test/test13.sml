datatype 'a t = T' of 'a * 'a list t

datatype 'a t = T' of ('a * 'a) t

datatype 'a t = T' of ('a * 'a * 'a * 'a * 'a) list t

datatype ('a, 'b) t = T' of (('a * 'a) list, 'b * 'b) t

(* This should work *)
datatype 'a t = T' of 'a t