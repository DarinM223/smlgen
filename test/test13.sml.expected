datatype 'a t = T' of 'a * 'a list t

datatype 'a t = T' of ('a * 'a) t

datatype 'a t = T' of ('a * 'a * 'a * 'a * 'a) list t

datatype ('a, 'b) t = T' of (('a * 'a) list, 'b * 'b) t

datatype 'a t = T' of {a: 'a} t

(* This should work *)
datatype 'a t =
  T' of ('a) t

(* This should also work *)
local
  val rec t = fn (t_1, a_) => fn T' t0 => "T' " ^ "(" ^ t_1 t0 ^ ")"
  val t = fn a_ =>
    let val rec t_1 = fn ? => t (t_1, a_) ? and t_0 = fn ? => t (t_1, a_) ?
    in t_0
    end
in val show = t
end
local
  val rec t = fn (t_1, a_) => fn (T' t0, T' t1) => t_1 (t0, t1)
  val t = fn a_ =>
    let val rec t_1 = fn ? => t (t_1, a_) ? and t_0 = fn ? => t (t_1, a_) ?
    in t_0
    end
in val compare = t
end
local
  val rec t = fn (t_1, a_) => fn (T' t0, T' t1) => t_1 (t0, t1)
  val t = fn a_ =>
    let val rec t_1 = fn ? => t (t_1, a_) ? and t_0 = fn ? => t (t_1, a_) ?
    in t_0
    end
in val op== = t
end
local
  val rec t = fn (t_1, a_) => fn T' t0 => t_1 t0
  val t = fn a_ =>
    let val rec t_1 = fn ? => t (t_1, a_) ? and t_0 = fn ? => t (t_1, a_) ?
    in t_0
    end
in val hash = t
end
datatype
  ( 'asuperextremelylongtypevariablethattakesupalotofcharacters
  , 'inordertotestthatthemaximumtypevariablesizeisnotbasedon
  , 'thestringlengthoftheprintextype
  , 'thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong
  ) t =
  T' of
    ( 'asuperextremelylongtypevariablethattakesupalotofcharacters
    , 'inordertotestthatthemaximumtypevariablesizeisnotbasedon
    , 'thestringlengthoftheprintextype
    , 'thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong
    ) t
local
  val rec t =
    fn ( t_0
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       , thestringlengthoftheprintextype_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , asuperextremelylongtypevariablethattakesupalotofcharacters_
       ) => fn T' t0 => "T' " ^ "(" ^ t_0 t0 ^ ")"
  val t =
    fn ( asuperextremelylongtypevariablethattakesupalotofcharacters_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , thestringlengthoftheprintextype_
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       ) =>
      let
        val rec t_0 = fn ? =>
          t
            ( t_0
            , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
            , thestringlengthoftheprintextype_
            , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
            , asuperextremelylongtypevariablethattakesupalotofcharacters_
            ) ?
      in
        t_0
      end
in val show = t
end
local
  val rec t =
    fn ( t_0
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       , thestringlengthoftheprintextype_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , asuperextremelylongtypevariablethattakesupalotofcharacters_
       ) => fn (T' t0, T' t1) => t_0 (t0, t1)
  val t =
    fn ( asuperextremelylongtypevariablethattakesupalotofcharacters_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , thestringlengthoftheprintextype_
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       ) =>
      let
        val rec t_0 = fn ? =>
          t
            ( t_0
            , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
            , thestringlengthoftheprintextype_
            , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
            , asuperextremelylongtypevariablethattakesupalotofcharacters_
            ) ?
      in
        t_0
      end
in val compare = t
end
local
  val rec t =
    fn ( t_0
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       , thestringlengthoftheprintextype_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , asuperextremelylongtypevariablethattakesupalotofcharacters_
       ) => fn (T' t0, T' t1) => t_0 (t0, t1)
  val t =
    fn ( asuperextremelylongtypevariablethattakesupalotofcharacters_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , thestringlengthoftheprintextype_
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       ) =>
      let
        val rec t_0 = fn ? =>
          t
            ( t_0
            , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
            , thestringlengthoftheprintextype_
            , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
            , asuperextremelylongtypevariablethattakesupalotofcharacters_
            ) ?
      in
        t_0
      end
in val op== = t
end
local
  val rec t =
    fn ( t_0
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       , thestringlengthoftheprintextype_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , asuperextremelylongtypevariablethattakesupalotofcharacters_
       ) => fn T' t0 => t_0 t0
  val t =
    fn ( asuperextremelylongtypevariablethattakesupalotofcharacters_
       , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
       , thestringlengthoftheprintextype_
       , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
       ) =>
      let
        val rec t_0 = fn ? =>
          t
            ( t_0
            , thiswouldnotworkotherwiseifthemaxtypevariablelengthisgreaterthanonehundredcharacterslong_
            , thestringlengthoftheprintextype_
            , inordertotestthatthemaximumtypevariablesizeisnotbasedon_
            , asuperextremelylongtypevariablethattakesupalotofcharacters_
            ) ?
      in
        t_0
      end
in val hash = t
end